### PR TITLE
[Feature-13763][K8S Task] collect real-time log

### LINKED ISSUE #13763: [Improvement][K8S Task] K8S task should collect whole pod log
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description
We should listen the pod log like https://github.com/apache/dolphinscheduler/issues/13751, rather just collect error logs when task runs failed.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Radeity**
## Purpose of the pull request
- This PR closes https://github.com/apache/dolphinscheduler/issues/13763

## Brief change log

- Use log watcher to collect real time, same with https://github.com/apache/dolphinscheduler/pull/13752
- Now, can only collect log of the first pod, have to consider in the future if one job launch multi pods.

## Verify this pull request
- Manually tested.
image

### CODE REVIEW

**Comment by Radeity** [comment]
Hi, @caishunfeng , would you like to help review this PR?

**Comment by Gallardot** [comment]
HI @Radeity , good job!
I tested this PR. Sometimes it's OK, but sometimes there's an error log. 
```
[INFO] 2023-07-10 19:42:57.004 +0800 - [K8sJobExecutor-echo-77] start to submit job
[INFO] 2023-07-10 19:42:57.073 +0800 - [K8sJobExecutor-echo-77] submitted job successfully
[INFO] 2023-07-10 19:42:57.095 +0800 - event received : job:echo-77 action:ADDED
[INFO] 2023-07-10 19:42:57.421 +0800 - event received : job:echo-77 action:MODIFIED
[INFO] 2023-07-10 19:42:57.421 +0800 - job echo-77 status 1
[INFO] 2023-07-10 19:43:02.459 +0800 - [K8S-pod-log] {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"container \"echo-77\" in pod \"echo-77-pngf7\" is waiting to start: ContainerCreating","reason":"BadRequest","code":400}
[INFO] 2023-07-10 19:43:25.109 +0800 - event received : job:echo-77 action:MODIFIED
[INFO] 2023-07-10 19:43:25.109 +0800 - job echo-77 status 0
[INFO] 2023-07-10 19:43:25.110 +0800 - [K8sJobExecutor-echo-77] succeed in k8s
```
Maybe some fault tolerance work is needed here. For example, before getting pod logs, check the status of the current task pod?

**Comment by Radeity** [comment]
Hi @Gallardot, thanks very much for your comment and test!  
> Maybe some fault tolerance work is needed here. For example, before getting pod logs, check the status of the current task pod?

Agree with you. Maybe we can print the error log with more specific error message like `Can not collect pod log, the status is {}` in this scenario, WDYT?

**Comment by Radeity** [comment]
Hi @Gallardot, thanks very much for your comment and test!  
> Maybe some fault tolerance work is needed here. For example, before getting pod logs, check the status of the current task pod?

Agree with you. I will add the status check to handle different pod status :D

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java [hunk 1]
```diff
@@ -181,15 +181,15 @@ private TaskExecutionStatus getApplicationStatus(KubernetesApplicationManagerCon
...
-                getDriverPod(kubernetesApplicationManagerContext);
-        List<Pod> driverPod = watchList.list().getItems();
-        if (CollectionUtils.isEmpty(driverPod)) {
+                getListenPod(kubernetesApplicationManagerContext);
+        List<Pod> podList = watchList.list().getItems();
+        if (CollectionUtils.isEmpty(podList)) {
             return null;
         }
-        Pod driver = driverPod.get(0);
+        Pod pod = podList.get(0);
```
Comment (Gallardot) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java]: Not sure if there will be multiple Pods. Maybe you should filter the latest pods based on when they were created?


https://github.com/apache/dolphinscheduler/blob/dev/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java#L100-L110

When getting the pod list, maybe add jobname as a label filter?

Comment (Radeity) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java]: > Not sure if there will be multiple Pods. Maybe you should filter the latest pods based on when they were created?

You are right, one job can launch multiple pods. For this scenario, I think it's better support to read logs from different pods in future PR, rather read from the latest one, WDYT?

> When getting the pod list, maybe add jobname as a label filter?

The label value here is `taskAppId` which can behave like a unique identifier of one job.

Comment (Gallardot) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java]: Got it. After the current PR merge, we create a new issue to track the issue

**Comment by Radeity** [comment]
Hi, @Gallardot , I have added a check only for **PENDING** pod, do you have some other suggestions here?
https://github.com/apache/dolphinscheduler/blob/2f9d91527af6c662f9bcb1ba17ece7c1d5c950b2/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java#L195-L199

**Comment by Gallardot** [comment]
> Hi, @Gallardot , I have added a check only for **PENDING** pod, do you have some other suggestions here?
> 
> https://github.com/apache/dolphinscheduler/blob/2f9d91527af6c662f9bcb1ba17ece7c1d5c950b2/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java#L195-L199

I think we can also add `UNKNOWN`.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java [hunk 2]
```diff
@@ -180,16 +182,25 @@ private TaskExecutionStatus getApplicationStatus(KubernetesApplicationManagerCon
...
+            } else {
+                podIsReady = true;
+            }
         }
-        Pod driver = driverPod.get(0);
 
-        return client.pods().inNamespace(driver.getMetadata().getNamespace())
-                .withName(driver.getMetadata().getName())
+        return client.pods().inNamespace(pod.getMetadata().getNamespace())
+                .withName(pod.getMetadata().getName())
```
Comment (Gallardot) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java:214]: Add inContainer() to further specify the task container for the pod. In some environments, the infrastructure may automatically add sidecar containers. If the container is not specified, an exception occurs.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java [hunk 3]
```diff
@@ -180,16 +182,25 @@ private TaskExecutionStatus getApplicationStatus(KubernetesApplicationManagerCon
...
-        if (CollectionUtils.isEmpty(driverPod)) {
-            return null;
+        boolean podIsReady = false;
+        Pod pod = null;
+        while (!podIsReady) {
+            FilterWatchListDeletable<Pod, PodList, PodResource> watchList =
+                    getListenPod(kubernetesApplicationManagerContext);
+            List<Pod> podList = watchList.list().getItems();
+            if (CollectionUtils.isEmpty(podList)) {
+                return null;
```
Comment (Gallardot) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java:202]: Should not return directly, perhaps the k8s controller manager is currently under high system load, and the scheduling is a little slow.

We should wait and retry, but of course we should set a reasonable number of retries.

**Review by Gallardot — CHANGES_REQUESTED** [review]
Rest LGTM

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java (hunk 3 above)
Comment (Radeity) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java:202]: Really appreciate, I'll make some changes later today :D

**Comment by Radeity** [comment]
Hi, @Gallardot , I've made some changes based on your suggestions, please help review :D

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/AbstractCommandExecutor.java [hunk 4]
```diff
@@ -327,7 +327,7 @@ private void collectPodLogIfNeeded() {
             ThreadUtils.sleep(SLEEP_TIME_MILLIS * 5L);
             try (
                     LogWatch watcher = ProcessUtils.getPodLogWatcher(taskRequest.getK8sTaskExecutionContext(),
-                            taskRequest.getTaskAppId())) {
+                            taskRequest.getTaskAppId(), "")) {
```
Comment (Gallardot) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/AbstractCommandExecutor.java:275]: Why is it an empty string?

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java [hunk 5]
```diff
@@ -97,16 +101,20 @@ public ResourceManagerType getResourceManagerType() {
...
-            log.warn("Expected driver pod 1, but get {}.", podList.size());
+        List<Pod> podList = null;
+        FilterWatchListDeletable<Pod, PodList, PodResource> watchList = null;
+        int retryTimes = 0;
+        while (CollectionUtils.isEmpty(podList) && retryTimes < MAX_RETRY_TIMES) {
+            watchList = client.pods()
+                    .inNamespace(kubernetesApplicationManagerContext.getK8sTaskExecutionContext().getNamespace())
+                    .withLabel(UNIQUE_LABEL_NAME, labelValue);
+            podList = watchList.list().getItems();
+            retryTimes += 1;
```
Comment (Gallardot) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java:119]: LGTM, while adding `thread.sleep` might be better.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/AbstractCommandExecutor.java (hunk 4 above)
Comment (Radeity) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/AbstractCommandExecutor.java:275]: In `AbstractCommandExecutor`, only need to get log watcher for Spark on K8S task, and we directly collect logs from driver pod, however, we can not set container name during task launching. Thus, we set empty string and the effect is same as collecting logs from all containers.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java (hunk 5 above)
Comment (Radeity) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java:119]: Oh sorry, I forgot to add this, thanks for your reminder!

**Comment by Radeity** [comment]
Hi, @caishunfeng @SbloodyS , would you guys like to help double check?

**Review by zhongjiajie — APPROVED** [review]

**Comment by zhongjiajie** [comment]
re-run the failed ci to find due to code bug or flaky test

**Review by SbloodyS — APPROVED** [review]
+1

**Comment by zhongjiajie** [comment]
seems it still when I re run the CI, could you please take a look @Radeity

**Pull request merged**